pdns/pdnsutil.cc

```diff
@@ -68,6 +68,519 @@ namespace {
+  // for indent
+  const std::string_view help;
+};
+
+// The commands entries are in a std::map, rather than std::unordered_map, in
+// order to be able to output them in sorted order, when listing the commands
+// in help displays.
+// The first element of the pair describes the group category.
+using groupCommandDispatcher = std::pair<std::string_view, std::map<std::string_view, commandEntry>>;
+
+// clang-format off [
+
+// AUTOPRIMARY
+
+static const groupCommandDispatcher autoprimaryCommands{
+  "Autoprimary",
+  {{"add", {true, addAutoprimary,
+    "IP NAMESERVER [ACCOUNT]",
+    "\tAdd a new autoprimary "}},
+   {"list", {true, listAutoprimaries,
+    "",
+    "\tList all autoprimaries"}},
+   {"remove", {true, removeAutoprimary,
+    "IP NAMESERVER",
+    "\tRemove an autoprimary"}}}
+};
+
+// CATALOG
+
+static const groupCommandDispatcher catalogCommands{
+  "Catalog Zone",
+  {{"list-members", {true, listMemberZones,
+    "CATALOG",
+    "\tList all members of catalog zone CATALOG"}},
+   {"set", {true, setCatalog,
+    "ZONE [CATALOG]",
+    "\tChange the catalog of ZONE to CATALOG, or removes ZONE from its current\n"
+    "\tcatalog if no catalog provided"}}}
+};
+
+// HSM
+
+#ifdef HAVE_P11KIT1 // [
+static const groupCommandDispatcher HSMCommands{
+  "HSM",
+  {{"assign", {true, HSMAssign,
+     "ZONE ALGORITHM {ksk|zsk} MODULE SLOT PIN LABEL [PUBLABEL]",
+     "\tAssign a Hardware Signing Module to a ZONE"}},
+   {"create-key", {true, HSMCreateKey,
+     "ZONE KEY_ID [BITS]",
+     "\tcreate a key using Hardware Signing Module for ZONE (use assign first);\n"
+     "\tBITS defaults to 2048"}}}
+};
+#endif // ]
+
+// META/МETADATA
+
+static const groupCommandDispatcher metadataCommands{
+  "Zone Metadata",
+  {{"add", {true, addMeta,
+    "ZONE KIND VALUE [VALUE...]",
+    "\tAdd zone metadata, this adds to the existing KIND"}},
+   {"get", {true, getMeta,
+    "ZONE [KIND...]",
+    "\tGet zone metadata. If no KIND is given, lists all known"}},
+   {"set", {true, setMeta,
+    "ZONE KIND [VALUE...]",
+    "\tSet zone metadata, replacing all existing records of KIND, optionally\n"
+    "\tproviding a value. An omitted value clears the metadata"}}}
+};
+
+// NETWORKS (VIEWS CONTEXT)
+
+static const groupCommandDispatcher networkCommands{
+  "Networks",
+  {{"list", {true, listNetwork,
+    "",
+    "\tList all defined networks with their chosen views"}},
+   {"set", {true, setNetwork,
+    "NET [VIEW]",
+    "\tSet the view for a network, or delete if no view argument."}}}
+};
+
+// RECORD/RRSET
+
+static const groupCommandDispatcher rrsetCommands{
+  "Zone Record",
+  {{"add", {true, addRecord,
+    R"(ZONE NAME TYPE [TTL] "CONTENT" ["CONTENT"...])",
+    "\tAdd one or more records to ZONE"}},
+   {"delete", {true, deleteRRSet,
+    "ZONE NAME TYPE",
+    "\tDelete named RRSET from ZONE"}},
+   {"hash", {true, hashZoneRecord,
+    "ZONE NAME",
+    "\tCalculate the NSEC3 hash for NAME in ZONE"}},
```

Comment:
this is not strictly rrset either, but I'm not sure we have a better spot for it